### PR TITLE
Resolve #2917: Allow converting a single/multi target indexing session to mutual one

### LINKED ISSUE #2917: Allow converting a single/multi target indexing session to mutual one
Right now, if an indexing session is very slow, the only way to convert it to mutual (and speed it up) is to disable the index and lose any progress that was already done.

### PR DESCRIPTION
**Pull request by jjezra**
This change should allow speeding up an indexing session with insufficient performance by converting it to mutual session without loosing all the progress that was already made.

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java [hunk 1]
```diff
@@ -1338,22 +1342,29 @@ public boolean shouldAllowUniquePendingState(FDBRecordStore store) {
...
+                    return isTypeConversionAllowed(TakeoverTypes.BY_RECORDS_TO_MUTUAL);
+                }
+            }
+            return false;
+        }
+
+        private boolean isTypeConversionAllowed(TakeoverTypes takeoverType) {
+            return
+                    allowedTakeoverSet == null ?
+                    allowTakeoverContinue :
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1366]: Perhaps it would make sense to push this into the builder, and have:
```
            public Builder allowTakeoverContinue(boolean allow) {
                this.allowedTakeoverSet = allow ? EnumSet.allOf(TakeoverTypes.class) : EnumSet.noneOf(TakeoverTypes.class);
                return this;
            }
```
and change the default value of `this.allowedTakeoverSet` to `EnumSet.noneOf(TakeoverTypes.class)` instead of `null`

Then this method just becomes `allowedTakeoverSet.contains(takeoverType)`, and may make sense to inline.

Perhaps better as a followup PR though.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java [hunk 2]
```diff
@@ -1338,22 +1342,29 @@ public boolean shouldAllowUniquePendingState(FDBRecordStore store) {
...
                                                          IndexBuildProto.IndexBuildIndexingStamp.Method oldMethod) {
-            if (allowedTakeoverSet != null) {
-                if (newMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
-                    if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MULTI_TARGET_BY_RECORDS) {
-                        return allowedTakeoverSet.contains(TakeoverTypes.MULTI_TARGET_TO_SINGLE);
-                    }
-                    if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS) {
-                        return allowedTakeoverSet.contains(TakeoverTypes.MUTUAL_TO_SINGLE);
-                    }
+            if (newMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1345]: So the total set of methods is:
```
  enum Method {
    INVALID      = 0;  // let zero be invalid
    BY_RECORDS   = 1;  // indexing by records - scanning all records in a record store
    BY_INDEX     = 2;  // indexing by a source index. When source index points to all possible records of the target index.
    SCRUB_REPAIR = 3;  // while readable, verify index validity. Repair if needed.
    MULTI_TARGET_BY_RECORDS = 4; // scan records, build multiple target indexes
    MUTUAL_BY_RECORDS = 5; // scan records, build multiple target indexes, while allowing multiple indexer processes
    NONE = 6; // return this stamp in query when null - never to be written in the DB
  };

```

And it would never make sense to go to, or from `INVALID` or `SCRUB_REPAIR`.
`BY_INDEX` to any of the others I think is not possible because the ranges would be wrong.
`NONE` to anything is implicitly supported (to the extent that `NONE` is a real type)

And the only other one that is not covered after this change `BY_RECORDS` to `MULTI_TARGET_BY_RECORDS`, although you could cause that to happen by first switching to `MUTUAL_BY_RECORDS`.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java [hunk 3]
```diff
@@ -1338,22 +1342,29 @@ public boolean shouldAllowUniquePendingState(FDBRecordStore store) {
...
+                if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MULTI_TARGET_BY_RECORDS) {
+                    return isTypeConversionAllowed(TakeoverTypes.MULTI_TARGET_TO_SINGLE);
+                }
+                if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS) {
+                    return isTypeConversionAllowed(TakeoverTypes.MUTUAL_TO_SINGLE);
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1350]: I know this existed before, but the comment below makes me wonder what happens if you have a `MUTUAL_BY_RECORDS` that has multiple targets? Does it get split up?

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java [hunk 4]
```diff
@@ -948,4 +948,61 @@ void testForbidConversionOfActiveMultiTarget() throws InterruptedException {
...
+                indexBuilder.buildIndex();
+            }
+        });
+        t1.start();
+        startBuildingSemaphore.acquire();
+        startBuildingSemaphore.release();
+        // Try as mutual and fail
+        try (OnlineIndexer indexBuilder = newIndexerBuilder(indexes)
+                .setIndexingPolicy(OnlineIndexer.IndexingPolicy.newBuilder()
+                        .setMutualIndexing(true))
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:998]: The key thing here is that it doesn't configure the policy to allow `BY_RECORDS_TO_MUTUAL`, right? Otherwise this indexer would keep building, and the other would fail.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 5]
```diff
@@ -403,11 +403,15 @@ void testMutualIndexingCrashAndRefuseContinueNonMutually() {
         assertAllValidated(indexes);
     }
 
-    OnlineIndexer.IndexingPolicy.Builder mutualTakeOverIndexingPolicy(boolean explicit) {
+    OnlineIndexer.IndexingPolicy.Builder mutualTakeOverIndexingPolicy(boolean explicit, boolean mutual) {
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:406]: Nit:
```suggestion
    OnlineIndexer.IndexingPolicy.Builder mutualTakeOverIndexingPolicy(boolean explicit, boolean toMutual) {
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 6]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
         assertAllValidated(indexes);
     }
 
+    @ParameterizedTest
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:458]: It seems like there should be a test where there are two indexers,
1. building `MULTI_TARGET_BY_RECORDS` and has `MUTUAL_TO_SINGLE` allowed
2. building `MUTUAL_BY_RECORDS` and has `BY_RECORDS_TO_MUTUAL` allowed

I believe the two will both do indexing, with the state constantly swapping.
If that is the case, probably worth calling out in the documentation, because you basically have no isolation between the two, and would presumably get a lot of conflicts if you had the two running.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 7]
```diff
@@ -403,11 +403,15 @@ void testMutualIndexingCrashAndRefuseContinueNonMutually() {
         assertAllValidated(indexes);
     }
 
-    OnlineIndexer.IndexingPolicy.Builder mutualTakeOverIndexingPolicy(boolean explicit) {
+    OnlineIndexer.IndexingPolicy.Builder mutualTakeOverIndexingPolicy(boolean explicit, boolean mutual) {
         final OnlineIndexer.IndexingPolicy.Builder builder = OnlineIndexer.IndexingPolicy.newBuilder();
+        final List<OnlineIndexer.IndexingPolicy.TakeoverTypes> conversionSet =
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:408]: I think it should be a separate PR, if at all, but it kind of seems like there should be a separate test class for takeover, rather than putting the tests in classes focused on the source or target.

That might also make it easier, or more obvious where some of this code could be de-duplicated and it could be easier to cover all the permutations that are allowed or not-allowed.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 8]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+                    throw new RecordCoreException(testThrowMsg);
+                })
+                .build()) {
+            RecordCoreException e = assertThrows(RecordCoreException.class, indexBuilder::buildIndex);
+            assertTrue(e.getMessage().contains(testThrowMsg));
+        }
+
+        // Continue as mutual
+        final FDBStoreTimer timer = new FDBStoreTimer();
+        try (OnlineIndexer indexBuilder = newIndexerBuilder(indexes, timer)
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:489]: Since the goal of this change is to avoid losing the work done by the first indexer, it seems like this test should assert that the mutual indexing has to index some records, but not all the records.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 9]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+        assertAllValidated(indexes);
+    }
+
+    @ParameterizedTest
+    @BooleanSource
+    void testSingleTargetContinueAsMutual(boolean explicit) {
+        // Start building with a single target, stop, continue as mutual
+        // Make sure that the mutual indexing is unblocked
+        // Note that a single target session produces a different indexing type stamp than a multi target session
+        final Index index = new Index("indexA", field("num_value_2"), EmptyKeyExpression.EMPTY, IndexTypes.VALUE, IndexOptions.UNIQUE_OPTIONS);
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:507]: It might be worth extracting a helper method since, I think the only difference in the test code between this and `testMultiTargetContinueAsMutual` is the `indexes` list.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 10]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+                .setConfigLoader(old -> {
+                    throw new RecordCoreException(testThrowMsg);
+                })
+                .build()) {
+            RecordCoreException e = assertThrows(RecordCoreException.class, indexBuilder::buildIndex);
+            assertTrue(e.getMessage().contains(testThrowMsg));
+        }
+
+        // Continue as mutual
+        try (OnlineIndexer indexBuilder = newIndexerBuilder(index)
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:530]: Similarly, this should assert about the number of records scanned.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 11]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+            indexBuilder.buildIndex(true);
+        }
+
+        // validate
+        assertReadable(indexes);
+        assertAllValidated(indexes);
+    }
+
+    @ParameterizedTest
+    @BooleanSource
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:543]: This doesn't use the parameter
```suggestion
    @Test
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 12]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+        indexes.add(new Index("indexA", field("num_value_2"), EmptyKeyExpression.EMPTY, IndexTypes.VALUE, IndexOptions.UNIQUE_OPTIONS));
+
+        int numRecords = 32;
+        populateData(numRecords);
+
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = allIndexesHook(indexes);
+        openSimpleMetaData(hook);
+        disableAll(indexes);
+
+        // Build as mutual, crash half way
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:559]:
```suggestion
        // Build as multi-target, crash half way
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 13]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+        List<Index> indexes = List.of(index);
+
+        int numRecords = 32;
+        populateData(numRecords);
+
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = allIndexesHook(indexes);
+        openSimpleMetaData(hook);
+        disableAll(indexes);
+
+        // Build as mutual, crash half way
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:517]:
```suggestion
        // Build as single-target, crash half way
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 14]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+        indexes.add(new Index("indexA", field("num_value_2"), EmptyKeyExpression.EMPTY, IndexTypes.VALUE, IndexOptions.UNIQUE_OPTIONS));
+
+        int numRecords = 32;
+        populateData(numRecords);
+
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = allIndexesHook(indexes);
+        openSimpleMetaData(hook);
+        disableAll(indexes);
+
+        // Build as mutual, crash half way
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:475]:
```suggestion
        // Build as multi-target, crash half way
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 15]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
         assertAllValidated(indexes);
     }
 
+    @ParameterizedTest
+    @BooleanSource
+    void testMultiTargetContinueAsMutual(boolean explicit) {
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:460]: It seems like there should also be tests of multi to-and-from mutual where they have different, potentially overlapping index sets.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 6 above)
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:458]: It seems like there should be a test exemplifying the behavior around SynchronizedSession, and how the mutual indexing doesn't respect it.
I believe if you use SynchronizedSession, and try to takeover from MULTI to SINGLE, while the former is still active due to the SynchronizedSession, but mutual will not do that, it will takeover even if the MULTI just committed a second ago. 

Perhaps this should be broken out into a separate body of work to improve the SynchronizedSession relationship with MUTUAL, but it might be pretty easy to add those tests, and then reference an issue to improve the situation.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java [hunk 16]
```diff
@@ -948,4 +948,61 @@ void testForbidConversionOfActiveMultiTarget() throws InterruptedException {
         // happy indexes assertion
         assertReadable(indexes);
     }
+
+    @Test
+    void testForbidConversionOfActiveMultiTargetToMutual() throws InterruptedException {
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:953]: Does the "Active" matter, wouldn't this code forbid the conversion regardless.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 17]
```diff
@@ -451,6 +455,132 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = allIndexesHook(indexes);
+        openSimpleMetaData(hook);
+        disableAll(indexes);
+
+        // Build as mutual, crash half way
+        final String testThrowMsg = "Intentionally crash during test";
+        try (OnlineIndexer indexBuilder = newIndexerBuilder(indexes)
+                .setLimit(10)
+                .setConfigLoader(old -> {
+                    throw new RecordCoreException(testThrowMsg);
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:564]: This fails and then forbids converting to MUTUAL, should there be another test where it pauses, and continues, similar to: `testForbidConversionOfActiveMultiTargetToMutual` above.
Should there be a test similar to this one but forbidding mutual_to_multi?

**Review by ScottDugas — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java (hunk 1 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1366]: ✅ 

(It was an easy change)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java (hunk 2 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1345]: As you've said, switching from `BY_INDEX` to any by-records method should never be allowed. 
Switching from `BY_RECORDS` to `MULTI_TARGET_BY_RECORDS` could have been allowed only if no records were scanned. 
It can make sense to verify that when switching from `BY_RECORDS` or `MULTI_TARGET_BY_RECORDS` to `MUTUAL_BY_RECORDS`, the very same target index set is being used. I'll try to add it.

Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1345]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java (hunk 3 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1350]: Yes. converting to `BY_RECORDS` will split any multi target. The original motivation was a fallback. 
I will add a comment to clarify.

Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1350]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java (hunk 4 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:998]: Yes. The test verifies that: 
1. The conversion is forbidden. 
2. The conversion attempt does not affect the original multi-target session, and that the multi-target session is completed successfully.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 5 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:406]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 6 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:458]: (Did you mean the first item to be 
1. building `BY_RECORDS` and has `MUTUAL_TO_SINGLE` allowed
?)

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 8 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:489]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 9 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:507]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 10 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:530]: (Done in the helper function)

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 11 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:543]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 12 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:559]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 13 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:517]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 14 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:475]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java (hunk 16 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:953]: Not really. The mutual does not use the sync lock.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 17 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java]: Do you mean build as mutual, pause, fail to convert to multi, then continue the mutual? 
If so: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 6 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:463]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 15 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:465]: ✅

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java [hunk 18]
```diff
@@ -1351,20 +1351,29 @@ public boolean shouldAllowTypeConversionContinue(IndexBuildProto.IndexBuildIndex
...
+                    // b) The mutual indexing continues only one target index (splitting it to one at a tiem)
+                    return newStamp.getTargetIndexCount() == 1 ||
+                            (newStamp.getTargetIndexCount() == oldStamp.getTargetIndexCount() &&
+                                    new HashSet<>(newStamp.getTargetIndexList()).containsAll(oldStamp.getTargetIndexList()));
+                }
+
+                if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
                     // Note that single-target-mutual is a private case of multi-target-mutual
-                    return isTypeConversionAllowed(TakeoverTypes.BY_RECORDS_TO_MUTUAL);
+                    return newStamp.getTargetIndexCount() == 1; // No real need to check the stamp's index name
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1369]: What happens if the indexes are different?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java [hunk 19]
```diff
@@ -1584,13 +1592,13 @@ public Builder allowTakeoverContinue() {
              * <li>"Single index by records" may continue a multi index session.</li>
              * <li>"Single index by records" may continue a mutually built session.</li>
              *  </ul>
-             *  This function may be too general; {@link #allowTakeoverContinue(Collection)} supports allowing a subset of these conversions.
+             *  This function will allow/disallow the all possible conversion types. {@link #allowTakeoverContinue(Collection)} supports allowing a subset of these conversions.
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1595]:
```suggestion
             *  This function will allow/disallow all possible conversion types. {@link #allowTakeoverContinue(Collection)} supports allowing a subset of these conversions.
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java [hunk 20]
```diff
@@ -910,19 +911,24 @@ void testForbidConversionOfActiveMultiTarget() throws InterruptedException {
...
-                        } catch (InterruptedException e) {
-                            throw new RuntimeException(e);
+                        if (passed.get()) {
+                            try {
+                                startBuildingSemaphore.release();
+                                pauseMutualBuildSemaphore.acquire(); // pause to try building indexes
+                            } catch (InterruptedException e) {
+                                throw new RuntimeException(e);
+                            }
+                            pauseMutualBuildSemaphore.release();
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:928]: Should this be in a `finally`?

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java [hunk 21]
```diff
@@ -970,19 +976,24 @@ void testForbidConversionOfActiveMultiTargetToMutual() throws InterruptedExcepti
...
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:993]: should this be in a `finally`?

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 22]
```diff
@@ -556,29 +585,256 @@ void testMultiTargetForbidContinueAsMutual() {
...
+        // Take over as mutual
+        try (OnlineIndexer indexBuilder = newIndexerBuilder()
+                .setTargetIndexes(indexes)
+                .setIndexingPolicy(mutualTakeOverIndexingPolicy(false, true)
+                        .setMutualIndexing())
+                .build()) {
+            indexBuilder.buildIndex();
         }
+
+        // let the other thread finish, verify that it did was not completed
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:769]:
```suggestion
        // let the other thread finish, verify that it was not completed
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 23]
```diff
@@ -440,7 +448,8 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
             try (OnlineIndexer indexBuilder = newIndexerBuilder()
                     .setIndex(index)
                     .setTimer(timer)
-                    .setIndexingPolicy(mutualTakeOverIndexingPolicy(explicit))
+                    .setIndexingPolicy(mutualTakeOverIndexingPolicy(explicit, false)
+                            .allowTakeoverContinue(true))
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:452]: doesn't `allowTakeover` here override the usage of `explicit` in the line above?
Wouldn't this be the same as:
```
                    .setIndexingPolicy(mutualTakeOverIndexingPolicy(false, false))
```

But I think the original version was intentional, to make sure that both `allowTakeoverContinue(true)` and `allowTakeoverContinue(List.of(...))` work the same.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java [hunk 24]
```diff
@@ -451,6 +460,383 @@ void testMutualIndexingCrashAndAllowContinueNonMutually(boolean explicit) {
...
+                    }
+                    passed.set(true);
+                    return old;
+                })
+                .build()) {
+            RecordCoreException e = assertThrows(RecordCoreException.class, indexBuilder::buildIndex);
+            assertTrue(e.getMessage().contains(testThrowMsg));
+        }
+
+        // Continue as mutual, by a different sublists of the indexes. All should fail
```
Comment (ScottDugas) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:605]: Should there be a similar test for going `mutual` to `multi-target`?

**Review by ScottDugas — CHANGES_REQUESTED** [review]
Mostly just typos, but one change looks suspicious.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java (hunk 18 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1369]: The old stamp does not have a target index (sort of backward compatibility, maybe it should be changed in the future). 
The new stamp is generated by the mutual indexer. If we check the stamp's validity here, we should check it everywhere.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java (hunk 19 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java:1595]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java (hunk 20 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:928]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java (hunk 21 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java:993]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 22 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:769]: ✅

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java (hunk 23 above)
Comment (jjezra) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMutualTest.java:452]: It is a leftover from a code change that was undone (but not fully undone - sorry). I removed line 452. 
✅

**Review by ScottDugas — APPROVED** [review]

**Review by ScottDugas — APPROVED** [review]
